tooling/nargo_cli/src/cli/compile_cmd.rs

```diff
@@ -54,28 +62,77 @@ pub(crate) fn run(
         selection,
         Some(NOIR_ARTIFACT_VERSION_STRING.to_owned()),
     )?;
-    let circuit_dir = workspace.target_directory_path();
 
+    if args.compile_options.expression_width.is_none() {
+        args.compile_options.expression_width = Some(backend.get_backend_info_or_default());
+    };
+
+    if args.watch {
+        watch_workspace(&workspace, &args.compile_options)
+            .map_err(|err| CliError::Generic(err.to_string()))?;
+    } else {
+        compile_workspace_full(&workspace, &args.compile_options)?;
+    }
+
+    Ok(())
+}
+
+fn watch_workspace(workspace: &Workspace, compile_options: &CompileOptions) -> notify::Result<()> {
+    let (tx, rx) = std::sync::mpsc::channel();
+
+    // No specific tickrate, max debounce time 1 seconds
+    let mut debouncer = new_debouncer(Duration::from_secs(1), None, tx)?;
+
+    // Add a path to be watched. All files and directories at that path and
+    // below will be monitored for changes.
+    debouncer.watcher().watch(&workspace.root_dir, RecursiveMode::Recursive)?;
+
+    let mut screen = std::io::stdout();
+    write!(screen, "{}", termion::cursor::Save).unwrap();
+    screen.flush().unwrap();
+    let _ = compile_workspace_full(workspace, compile_options);
+    for res in rx {
+        let debounced_events = res.map_err(|mut err| err.remove(0))?;
```

Comment:
even though this debounces events, it appears that multiple modifications could result in duplicated recompilation?

if so, could add a `last_compile_at: Instant` to filter out old events:
```rust
let mut last_compile_at = Instant::now();
let _ = compile_workspace_full(..);
for res in rx {
  ..
  let noir_files_modified = debounced_events.iter().filter(|event| event.time > last_compile_at).any(  
```

Reply:
I'm not clear on the situation that would result in extra redundant compilations here. Can you go into more detail?

Reply:
Looking closer, I think the cases I was thinking of are covered by `new_debouncer(Duration::from_secs(1), ..`

There are a couple cases where multiple events can be emitted within a short time, e.g. renaming from the CLI or some editors creating/replacing on save, but none of them exceed the 1-second window so I think we're set.